File: DecisionStump/py/DecisionStump.py
import random
def sign(n):
	if n <= 0:
		return -1
	else:
		return 1
def traindata(n):
	data = []
	for i in range(0,n):
		rand = random.uniform(-1,1)
		data.append(rand)
	return data

def flipSign(n):
	return -1*n

def getOutputData(input):
	output = []
	for i in input:
		rand = random.uniform(0,1)
		if rand <=0.2:
			output.append(flipSign(sign(i)))
		else:
			output.append(sign(i))
	return output

def getMinErrInSample(input,output,bound):
	minErrInSample = 1.0
	result = 0
	for i in list(range(19)):
		threshold = (input[i] + input[i+1])/2
		currentErrInSample = calErrInSample(input,output,bound,threshold)
		if currentErrInSample < minErrInSample:
			minErrInSample = currentErrInSample
			result = threshold
	return minErrInSample,bound,result

def getErrOutSample(bound,threshold):
	result = 0.5 + 0.3*bound*(abs(threshold-1))
	return result

def calErrInSample(input,output,bound,threshold):
	errCount = 0
	for i in list(range(20)):
		if output[i]!=bound*sign(input[i]-threshold):
			errCount = errCount + 1
	return errCount/20

def decisionstump(input,output):
	errPositive,bound,threshold = getMinErrInSample(input,output,1)
	errNegative,bound,threshold = getMinErrInSample(input,output,-1)
	if errPositive < errNegative:
		return errPositive,bound,threshold
	else:
		return errNegative,bound,threshold

errInTotal = 0
errOutTotal = 0
for i in range(5000):
	input = traindata(20)
	output = getOutputData(input)
	err,bound,threshold = decisionstump(input,output)
	errInTotal = errInTotal + err
	errOutTotal = errOutTotal + getErrOutSample(bound,threshold)
print('average in sample error:',errInTotal/5000)
print('average out sample error:',errOutTotal/5000)

File: DecisionStump/java/DecisionStump/src/DecisionStump.java
import java.util.Random;


public class DecisionStump {
	public static void main(String[] args){
		double errorInTotal = 0;
		double errorOutTotal = 0;
		double coefficient;
		double threshold;
		for (int i=0;i<5000;i++){
			double[] input = new double[20];
			double[] output = new double[20];
			double[] result = new double[3];
			input = getTrainingData(20);
			output = getOutput(input);
			result = decisionStump(input,output);
			errorInTotal = errorInTotal + result[0];
			errorOutTotal = errorOutTotal + calErrOut(result[1],result[2]);
		}
		System.out.println("the average in sample error:"+errorInTotal/5000);
		System.out.println("the average out sample error:"+errorOutTotal/5000);
	}
	public static double[] decisionStump(double[] input,double[] output){
		double[] errPositive = new double[3];
		double[] errNegative = new double[3];
		errPositive = getMinErrIn(input,output,1);
		errNegative = getMinErrIn(input,output,-1);
		if (errPositive[0] < errNegative[0]){
			return errPositive;
		} else
			return errNegative;
	}
	public static double sign (double x){
		if (x<=0)
			return -1.0;
		else
			return 1.0;
	}
	public static double[] getTrainingData(int n){
		double[] x = new double[n];
		Random random = new Random();
		for(int i=0;i<n;i++)
			x[i] = -1 + 2*random.nextDouble();
		return x;
	}
	public static double[] getOutput(double[] input){
		double[] y = new double[20];
		Random random = new Random();
		for (int i=0;i<20;i++){
			double rand = random.nextDouble();
			if (rand<=0.2)
				y[i] = -1*sign(input[i]);
			else
				y[i] = sign(input[i]);
		}
		return y;
	}
	public static double[] getMinErrIn(double[] input,double[] output,double s){
		double minErrIn = 1;
		double currErrIn;
		double coef;
		double threshold;
		double[] result = new double[3];
		for (int i=0;i<input.length-1;i++){
			coef = s;
			threshold = (input[i] + input[i+1])/2;
			currErrIn = calErrIn(input,output,coef,threshold);
			if (currErrIn < minErrIn){
				minErrIn = currErrIn;
				result[0] = minErrIn;
				result[1] = coef;
				result[2] = threshold;
			}
		}
		return result;
	}
	public static double calErrIn(double[] input,double[] output,double coef,double threshold){
		double errCount = 0;
		for (int i=0;i<input.length;i++){
			if (output[i]!=coef*sign(input[i]-threshold))
				errCount++;
		}
		return errCount/input.length;
	}
	public static double calErrOut(double coef,double threshold){
		return 0.5+0.3*(Math.abs(threshold)-1);
	}
}
